stable_plugins/visualization/file_types/qasm_visualization.py

```diff
@@ -172,47 +177,49 @@ def render(self, data: Mapping, errors: dict, valid: bool):
         )
 
 
+class ImageNotFinishedError(Exception):
+    pass
+
+
 @QASM_BLP.route("/circuits/")
 @QASM_BLP.response(HTTPStatus.OK, description="Circuit image.")
 @QASM_BLP.arguments(
-    QasmInputParametersSchema(
-        partial=True, unknown=EXCLUDE, validate_errors_as_result=False
-    ),
+    QasmInputParametersSchema(unknown=EXCLUDE),
     location="query",
     required=True,
 )
 @QASM_BLP.require_jwt("jwt", optional=True)
 def get_circuit_image(data: Mapping):
-    circuit_url = data.get("data", None)
-    if circuit_url is None:
+    url = data.get("data", None)
+    if not url:
         abort(HTTPStatus.BAD_REQUEST)
-    url_hash = hashlib.sha256(circuit_url.encode("utf-8")).hexdigest()
-    image_path = PluginState.get_value(
-        QasmVisualization.instance.identifier, url_hash, None
-    )
-    if image_path is None:
-        timeout = PluginState.get_value(
-            QasmVisualization.instance.identifier, f"timeout_{url_hash}", False
-        )
-        if not timeout:
-            start_task(circuit_url)
-            timeout = 2
-        else:
-            # increase timeout for next request
+    url_hash = hashlib.sha256(url.encode("utf-8")).hexdigest()
+    image = DataBlob.get_value(QasmVisualization.instance.identifier, url_hash, None)
+    if image is None:
+        response = Response("Image not yet created!", HTTPStatus.SERVICE_UNAVAILABLE)
+        response.headers["Retry-After"] = 5
+        if not PluginState.get_value(
```

Comment:
In the else case you could get an awaitable asyncResult from the task id. (see task api in plugin runner)